### PR TITLE
[Merged by Bors] - Add min version to join attribute

### PR DESCRIPTION
**Pull request by morenol**

### CODE REVIEW

File: crates/fluvio-smartstream-derive/src/generator/aggregate.rs [hunk 1]
```diff
@@ -50,7 +50,8 @@ pub fn generate_aggregate_smartstream(func: &SmartStreamFn, has_params: bool) ->
 
                 let input_data = Vec::from_raw_parts(ptr, len, len);
                 let mut smartstream_input = SmartStreamAggregateInput::default();
-                if let Err(_err) = Decoder::decode(&mut smartstream_input, &mut std::io::Cursor::new(input_data), 0) {
+                // 13 is version for aggregate
+                if let Err(_err) = Decoder::decode(&mut smartstream_input, &mut std::io::Cursor::new(input_data), 13) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/aggregate.rs:54]: can you change this to constant?

File: crates/fluvio-smartstream-derive/src/generator/aggregate.rs [hunk 2]
```diff
@@ -60,7 +61,7 @@ pub fn generate_aggregate_smartstream(func: &SmartStreamFn, has_params: bool) ->
 
                 let records_input = smartstream_input.base.record_data;
                 let mut records: Vec<Record> = vec![];
-                if let Err(_err) = Decoder::decode(&mut records, &mut std::io::Cursor::new(records_input), 0) {
+                if let Err(_err) = Decoder::decode(&mut records, &mut std::io::Cursor::new(records_input), 13) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/aggregate.rs:64]: use constant

File: crates/fluvio-smartstream-derive/src/generator/aggregate.rs [hunk 3]
```diff
@@ -95,7 +96,7 @@ pub fn generate_aggregate_smartstream(func: &SmartStreamFn, has_params: bool) ->
 
                 // ENCODING
                 let mut out = vec![];
-                if let Err(_) = Encoder::encode(&mut output, &mut out, 0) {
+                if let Err(_) = Encoder::encode(&mut output, &mut out, 13) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/aggregate.rs:99]: use constant

File: crates/fluvio-smartengine/src/smartstream/map.rs [hunk 4]
```diff
@@ -39,7 +39,7 @@ impl SmartStream for SmartStreamMap {
             return Err(internal_error.into());
         }
 
-        let output: SmartStreamOutput = self.base.read_output()?;
+        let output: SmartStreamOutput = self.base.read_output(0)?;
```
Comment (sehz) [inline crates/fluvio-smartengine/src/smartstream/map.rs:42]: use constant

File: crates/fluvio-smartstream-derive/src/generator/array_map.rs [hunk 5]
```diff
@@ -51,13 +51,13 @@ pub fn generate_array_map_smartstream(func: &SmartStreamFn, has_params: bool) ->
 
                 let input_data = Vec::from_raw_parts(ptr, len, len);
                 let mut smartstream_input = SmartStreamInput::default();
-                if let Err(_err) = Decoder::decode(&mut smartstream_input, &mut std::io::Cursor::new(input_data), 0) {
+                if let Err(_err) = Decoder::decode(&mut smartstream_input, &mut std::io::Cursor::new(input_data), 15) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/array_map.rs:54]: use constant

File: crates/fluvio-smartstream-derive/src/generator/array_map.rs [hunk 6]
```diff
@@ -51,13 +51,13 @@ pub fn generate_array_map_smartstream(func: &SmartStreamFn, has_params: bool) ->
...
                     return SmartStreamInternalError::DecodingBaseInput as i32;
                 }
 
                 let records_input = smartstream_input.record_data;
                 let mut records: Vec<Record> = vec![];
-                if let Err(_err) = Decoder::decode(&mut records, &mut std::io::Cursor::new(records_input), 0) {
+                if let Err(_err) = Decoder::decode(&mut records, &mut std::io::Cursor::new(records_input), 15) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/array_map.rs:60]: use constant

File: crates/fluvio-smartstream-derive/src/generator/array_map.rs [hunk 7]
```diff
@@ -95,7 +95,7 @@ pub fn generate_array_map_smartstream(func: &SmartStreamFn, has_params: bool) ->
 
                 // ENCODING
                 let mut out = vec![];
-                if let Err(_) = Encoder::encode(&mut output, &mut out, 0) {
+                if let Err(_) = Encoder::encode(&mut output, &mut out, 15) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/array_map.rs:98]: use constant

File: crates/fluvio-smartstream-derive/src/generator/filter_map.rs [hunk 8]
```diff
@@ -51,13 +51,13 @@ pub fn generate_filter_map_smartstream(func: &SmartStreamFn, has_params: bool) -
 
                 let input_data = Vec::from_raw_parts(ptr, len, len);
                 let mut smartstream_input = SmartStreamInput::default();
-                if let Err(_err) = Decoder::decode(&mut smartstream_input, &mut std::io::Cursor::new(input_data), 0) {
+                if let Err(_err) = Decoder::decode(&mut smartstream_input, &mut std::io::Cursor::new(input_data), 15) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/filter_map.rs:54]: use constant

File: crates/fluvio-smartstream-derive/src/generator/filter_map.rs [hunk 9]
```diff
@@ -51,13 +51,13 @@ pub fn generate_filter_map_smartstream(func: &SmartStreamFn, has_params: bool) -
...
                     return SmartStreamInternalError::DecodingBaseInput as i32;
                 }
 
                 let records_input = smartstream_input.record_data;
                 let mut records: Vec<Record> = vec![];
-                if let Err(_err) = Decoder::decode(&mut records, &mut std::io::Cursor::new(records_input), 0) {
+                if let Err(_err) = Decoder::decode(&mut records, &mut std::io::Cursor::new(records_input), 15) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/filter_map.rs:60]: use constant

File: crates/fluvio-smartstream-derive/src/generator/filter_map.rs [hunk 10]
```diff
@@ -93,7 +93,7 @@ pub fn generate_filter_map_smartstream(func: &SmartStreamFn, has_params: bool) -
 
                 // ENCODING
                 let mut out = vec![];
-                if let Err(_) = Encoder::encode(&mut output, &mut out, 0) {
+                if let Err(_) = Encoder::encode(&mut output, &mut out, 15) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/filter_map.rs:96]: use constant

File: crates/fluvio-smartstream-derive/src/generator/join.rs [hunk 11]
```diff
@@ -51,20 +51,20 @@ pub fn generate_join_smartstream(func: &SmartStreamFn, has_params: bool) -> Toke
 
                 let input_data = Vec::from_raw_parts(ptr, len, len);
                 let mut smartstream_input = SmartStreamInput::default();
-                if let Err(_err) = Decoder::decode(&mut smartstream_input, &mut std::io::Cursor::new(input_data), 0) {
+                if let Err(_err) = Decoder::decode(&mut smartstream_input, &mut std::io::Cursor::new(input_data), 17) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/join.rs:54]: use constant

File: crates/fluvio-smartstream-derive/src/generator/join.rs [hunk 12]
```diff
@@ -51,20 +51,20 @@ pub fn generate_join_smartstream(func: &SmartStreamFn, has_params: bool) -> Toke
...
                     return SmartStreamInternalError::DecodingBaseInput as i32;
                 }
 
                 let records_input = smartstream_input.record_data;
                 let mut records: Vec<Record> = vec![];
-                if let Err(_err) = Decoder::decode(&mut records, &mut std::io::Cursor::new(records_input), 0) {
+                if let Err(_err) = Decoder::decode(&mut records, &mut std::io::Cursor::new(records_input), 17) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/join.rs:60]: use constant

File: crates/fluvio-smartstream-derive/src/generator/join.rs [hunk 13]
```diff
@@ -51,20 +51,20 @@ pub fn generate_join_smartstream(func: &SmartStreamFn, has_params: bool) -> Toke
...
                     return SmartStreamInternalError::DecodingRecords as i32;
                 };
 
                 let join_last_record_input = smartstream_input.join_record;
                 let mut join_last_record: Option<Record> = None;
-                if let Err(_err) = Decoder::decode(&mut join_last_record, &mut std::io::Cursor::new(join_last_record_input), 0) {
-                    return SmartStreamInternalError::DecodingRecords as i32;
+                if let Err(_err) = Decoder::decode(&mut join_last_record, &mut std::io::Cursor::new(join_last_record_input), 17) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/join.rs:66]: use constant

File: crates/fluvio-smartstream-derive/src/generator/join.rs [hunk 14]
```diff
@@ -102,7 +102,7 @@ pub fn generate_join_smartstream(func: &SmartStreamFn, has_params: bool) -> Toke
 
                 // ENCODING
                 let mut out = vec![];
-                if let Err(_) = Encoder::encode(&mut output, &mut out, 0) {
+                if let Err(_) = Encoder::encode(&mut output, &mut out, 17) {
```
Comment (sehz) [inline crates/fluvio-smartstream-derive/src/generator/join.rs:105]: use constant

File: crates/fluvio-spu-schema/src/server/stream_fetch.rs [hunk 15]
```diff
@@ -75,7 +75,7 @@ where
     R: Debug + Decoder + Encoder,
 {
     const API_KEY: u16 = SpuServerApiKey::StreamFetch as u16;
-    const DEFAULT_API_VERSION: i16 = SMART_MODULE_API;
+    const DEFAULT_API_VERSION: i16 = JOIN_WASM_API;
```
Comment (sehz) [inline crates/fluvio-spu-schema/src/server/stream_fetch.rs:78]: since we didn't publish SMART_MODULE_API, we can still use same

**Review by sehz — CHANGES_REQUESTED** [review]

**Comment by morenol** [comment]
I reimported constants in fluvio_smartstream so I can use them in fluvio-smartstream-derive

**Comment by morenol** [comment]
I am using now WASM_MODULE_V2_API (12) for filter and map. It was using previously `0`

**Review by sehz — APPROVED** [review]
LGTM. Nice work

**Pull request closed without merging**